virttest/qemu_devices/qcontainer.py

```diff
@@ -2829,3 +2832,93 @@ def fs_define_by_params(self, name, params, bus=None):
         else:
             raise ValueError("unsupported filesystem driver type")
         return devices
+
+    def tpm_define_by_params(self, name, params):
+        """
+        Create TPM devices by params.
+
+        :param name: Object name of TPM device.
+        :type name: str
+        :param params: TPM device params.
+        :type params: dict
+        :return: List of corresponding TPM devices.
+        :rtype: list
+        """
+        model = params.get("tpm_model")
+        backend = params.get("tpm_backend")
+        version = params.get("tpm_version")
+
+        tmp_dir = data_dir.get_tmp_dir()
+
+        def _handle_log(line):
+            try:
+                utils_misc.log_line('%s_%s_swtpm_setup.log' % (self.vmname, name), line)
+            except Exception as e:
+                logging.warn("Can't log %s_%s_swtpm_setup output: %s.", self.vmname, name, e)
+
+        def _emulator_setup(binary, extra_options=None):
+            setup_cmd = binary
+            if version in ('2.0',):
+                setup_cmd += ' --tpm2'
+
+            tpm_path = os.path.join(tmp_dir, '%s_%s_tpm_state' % (self.vmname, name))
+            if os.path.exists(tpm_path):
+                shutil.rmtree(tpm_path, ignore_errors=True)
+            os.makedirs(tpm_path)
+            setup_cmd += " --tpm-state %s" % tpm_path
+
+            setup_cmd += (" --createek --create-ek-cert"
+                          " --create-platform-cert"
+                          " --lock-nvram --not-overwrite")
+
+            if extra_options:
+                setup_cmd += extra_options
+
+            logging.info('Running TPM emulator setup command: %s.', setup_cmd)
+            _process = aexpect.run_bg(setup_cmd, None, _handle_log, auto_close=False)
+            status_ending = 'Ending vTPM manufacturing'
+            _process.read_until_any_line_matches(status_ending, timeout=5)
+            return tpm_path
+
+        devices = []
+        if backend == 'emulator':
+            tpm_bin = params.get("tpm_bin", "/usr/bin/swtpm")
+            tpm_setup_bin = params.get("tpm_setup_bin", "/usr/bin/swtpm_setup")
+            tpm_bin_extra_options = params.get("tpm_bin_extra_options")
+            tpm_setup_bin_extra_options = params.get("tpm_setup_bin_extra_options")
+            sock_path = os.path.join(tmp_dir, '_'.join((self.vmname, name, 'swtpm.sock')))
+
+            storage_path = _emulator_setup(tpm_setup_bin, tpm_setup_bin_extra_options)
+            swtpmd = qdevices.QswtpmDev(name, tpm_bin, sock_path, storage_path,
+                                        version, tpm_bin_extra_options)
```

Comment:
So the step like `python3 -c "import socket as s; sock = s.socket(s.AF_UNIX); sock.bind('/tmp/guest-swtpm.sock')"` is not necessary, directly start the daemon with the socket path is okay, right?

Reply:
@qccheng-redhat Please help confirm it too. Thanks.

Reply:
Yes,  directly start the daemon with the socket path is okay.  I have updated the test case . Thanks.

Reply:
Okay, thanks @qcheng-redhat